File: src/commands/forums.ts
import { stripHtmlTags, getOutputFormat, formatTimestamp } from "../lib/utils.ts";
import { Command } from "commander";
import type { OutputFormat } from "../lib/types.ts";
import { getEnrolledCoursesApi, getForumsApi, getForumDiscussionsApi, getDiscussionPostsApi, addForumDiscussionApi, addForumPostApi, deleteForumPostApi, resolveForumId } from "../lib/moodle.ts";
import { createApiContext } from "../lib/auth.ts";
import { formatAndOutput } from "../index.ts";

interface ForumWithCourse {
  course_id: number;
  course_name: string;
  cmid: string;
  forum_id: number;
  name: string;
  intro: string;
  timemodified: number;
  // url: string;
}

export function registerForumsCommand(program: Command): void {
  const forumsCmd = program.command("forums");
  forumsCmd.description("Forum operations");

  async function listForums(classification?: "inprogress" | "past" | "future" | "all") {
    const apiContext = await createApiContext({});
    if (!apiContext) {
      process.exitCode = 1;
      return;
    }

    const courses = await getEnrolledCoursesApi(apiContext.session, {
      classification,
    });

    const courseIds = courses.map(c => c.id);
    const wsForums = await getForumsApi(apiContext.session, courseIds);

    const courseMap = new Map(courses.map(c => [c.id, c]));
    const allForums: ForumWithCourse[] = [];
    for (const wsForum of wsForums) {
      const course = courseMap.get(wsForum.courseid);
      if (course) {
        allForums.push({
          course_id: wsForum.courseid,
          course_name: course.fullname,
          intro: wsForum.intro,
          cmid: wsForum.cmid.toString(),
          forum_id: wsForum.id,
          name: wsForum.name,
          timemodified: wsForum.timemodified,
        });
      }
    }

    formatAndOutput(
      allForums as unknown as Record<string, unknown>[],
      "json",
      apiContext.log,
      { status: "success", timestamp: new Date().toISOString(), total_courses: courses.length, total_forums: allForums.length }
    );
  }

  forumsCmd
    .command("list")
    .description("List forums from in-progress courses")
    .action(() => listForums("inprogress"));

  forumsCmd
    .command("list-all")
    .description("List all forums across all courses")
    .option("--level <type>", "Course level: in_progress (default) | all", "in_progress")
    .action(async (options) => {
      const classification = options.level === "all" ? undefined : "inprogress";
      await listForums(classification);
    });

  forumsCmd
    .command("discussions")
    .description("List discussions in a forum (use forum ID)")
    .argument("<forum-id>", "Forum ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (forumId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const resolved = await resolveForumId(apiContext.session, forumId);
      if (!resolved) {
        apiContext.log.error("Forum not found");
        process.exitCode = 1;
        return;
      }

      const discussions = await getForumDiscussionsApi(apiContext.session, resolved.forumId);

      const items = discussions.map(d => ({
        id: d.id,
        name: d.name,
        user_id: d.userId,
        time_modified: d.timeModified,
        post_count: d.postCount,
        unread: d.unread,
        message: stripHtmlTags(d.message || ""),
      }));

      formatAndOutput(
        items as unknown as Record<string, unknown>[],
        output,
        apiContext.log,
        { status: "success", timestamp: new Date().toISOString(), forum_id: resolved.forumId, forum_name: resolved.name ?? null, course_id: resolved.courseid ?? null, total_discussions: discussions.length }
      );
    });

  forumsCmd
    .command("posts")
    .description("Show posts in a discussion")
    .argument("<discussion-id>", "Discussion ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (discussionId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const posts = await getDiscussionPostsApi(apiContext.session, parseInt(discussionId, 10));

      const items = posts.map(p => ({
        id: p.id,
        subject: p.subject,
        author: p.author,
        author_id: p.authorId,
        created: formatTimestamp(p.created),
        modified: formatTimestamp(p.modified),
        message: p.message,
        unread: p.unread,
      }));

      formatAndOutput(
        items as unknown as Record<string, unknown>[],
        output,
        apiContext.log,
        { status: "success", timestamp: new Date().toISOString(), discussion_id: discussionId, total_posts: posts.length }
      );
    });

  forumsCmd
    .command("post")
    .description("Post a new discussion to a forum")
    .argument("<forum-id>", "Forum ID")
    .argument("<subject>", "Discussion subject")
    .argument("<message>", "Discussion message")
    .option("--subscribe", "Subscribe to the discussion", false)
    .option("--pin", "Pin the discussion", false)
    .action(async (forumId, subject, message, options, command) => {
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const { log, session } = apiContext;

      const resolved = await resolveForumId(session, forumId);
      if (!resolved) {
        log.error(`Forum not found: ${forumId}`);
        process.exitCode = 1;
        return;
      }

      log.info(`Posting to forum: ${resolved.name ?? forumId}`);

      const result = await addForumDiscussionApi(
        session,
        resolved.forumId,
        subject,
        message
      );

      if (result.success) {
        log.success(`✓ Discussion posted successfully!`);
        log.info(`  Discussion ID: ${result.discussionId}`);
      } else {
        log.error(`✗ Failed to post discussion: ${result.error}`);
        process.exitCode = 1;
      }
    });

  forumsCmd
    .command("reply")
    .description("Reply to a discussion post")
    .argument("<post-id>", "Parent post ID to reply to")
    .argument("<subject>", "Reply subject")
    .argument("<message>", "Reply message")
    .option("--attachment-id <id>", "Draft file ID for attachment")
    .option("--inline-attachment-id <id>", "Draft file ID for inline attachment")
    .action(async (postId, subject, message, options, command) => {
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const { log, session } = apiContext;

      log.info(`Replying to post: ${postId}`);
      log.info(`  Subject: ${subject}`);
      log.info(`  Message: ${message}`);
      if (options.attachmentId) {
        log.info(`  Attachment ID: ${options.attachmentId}`);
      }

      const result = await addForumPostApi(
        session,
        parseInt(postId, 10),
        subject,
        message,
        {
          attachmentId: options.attachmentId ? parseInt(options.attachmentId, 10) : undefined,
          inlineAttachmentId: options.inlineAttachmentId ? parseInt(options.inlineAttachmentId, 10) : undefined,
        }
      );

      if (result.success) {
        log.success(`✓ Reply posted successfully!`);
        log.info(`  Post ID: ${result.postId}`);
      } else {
        log.error(`✗ Failed to post reply: ${result.error}`);
        process.exitCode = 1;
      }
    });

  forumsCmd
    .command("delete")
    .description("Delete a forum post or discussion (by post ID)")
    .argument("<post-id>", "Post ID to delete (deletes entire discussion if it's the first post)")
    .action(async (postId, options, command) => {
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const { log, session } = apiContext;

      const result = await deleteForumPostApi(session, parseInt(postId, 10));

      if (result.success) {
        log.success(`✓ Post ${postId} deleted successfully!`);
      } else {
        log.error(`✗ Failed to delete post: ${result.error}`);
        process.exitCode = 1;
      }
    });
}

File: src/commands/forums.rs
use anyhow::Result;
use crate::Cli;
use crate::moodle::course::get_enrolled_courses_api;
use crate::moodle::forum::{
    get_forums_api, get_forum_discussions_api, get_discussion_posts_api,
    add_discussion_api, add_discussion_post_api, delete_post_api,
};
use crate::output::format_and_output;
use crate::utils::{strip_html_tags, format_moodle_date};
use super::{ApiCtx, level_to_classification};

pub async fn run(cmd: &crate::ForumsCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli.config.as_ref(), cli.output, cli.verbose, cli.silent)?;

    match cmd {
        crate::ForumsCommands::List => {
            list_forums(&ctx, "inprogress").await?;
        }

        crate::ForumsCommands::ListAll { level } => {
            let classification = level_to_classification(*level);
            list_forums(&ctx, classification).await?;
        }

        crate::ForumsCommands::Discussions { forum_id } => {
            let discussions = get_forum_discussions_api(
                &ctx.client, &ctx.session, *forum_id, None, None, None, None,
            ).await?;

            let items: Vec<serde_json::Value> = discussions.iter().map(|d| serde_json::json!({
                "id": d.id,
                "name": d.name,
                "user_id": d.user_id,
                "time_modified": d.time_modified,
                "post_count": d.post_count,
                "unread": d.unread,
                "message": d.message.as_deref().map(strip_html_tags),
            })).collect();

            ctx.log.info(&format!("Forum {}: {} discussions", forum_id, items.len()));
            format_and_output(&items, ctx.output, None);
        }

        crate::ForumsCommands::Posts { discussion_id } => {
            let posts = get_discussion_posts_api(
                &ctx.client, &ctx.session, *discussion_id,
            ).await?;

            let items: Vec<serde_json::Value> = posts.iter().map(|p| serde_json::json!({
                "id": p.id,
                "subject": p.subject,
                "author": p.author,
                "author_id": p.author_id,
                "created": format_moodle_date(Some(p.created)),
                "modified": format_moodle_date(Some(p.modified)),
                "message": p.message,
                "unread": p.unread,
            })).collect();

            ctx.log.info(&format!("Discussion {}: {} posts", discussion_id, items.len()));
            format_and_output(&items, ctx.output, None);
        }

        crate::ForumsCommands::Post { forum_id, subject, message, .. } => {
            ctx.log.info(&format!("Posting to forum {}...", forum_id));
            match add_discussion_api(&ctx.client, &ctx.session, *forum_id, subject, message).await? {
                Some(discussion_id) => {
                    ctx.log.success("Discussion posted successfully!");
                    ctx.log.info(&format!("  Discussion ID: {}", discussion_id));
                    let result = serde_json::json!({
                        "action": "post",
                        "forum_id": *forum_id,
                        "discussion_id": discussion_id,
                        "success": true,
                    });
                    format_and_output(&[result], ctx.output, None);
                }
                None => {
                    anyhow::bail!("Post appeared to succeed but no discussion ID returned");
                }
            }
        }

        crate::ForumsCommands::Reply { post_id, subject, message, attachment_id, inline_attachment_id } => {
            ctx.log.info(&format!("Replying to post {}...", post_id));
            match add_discussion_post_api(
                &ctx.client, &ctx.session, *post_id, subject, message,
                *inline_attachment_id, *attachment_id,
            ).await? {
                Some(new_post_id) => {
                    ctx.log.success("Reply posted successfully!");
                    ctx.log.info(&format!("  Post ID: {}", new_post_id));
                    let result = serde_json::json!({
                        "action": "reply",
                        "post_id": *post_id,
                        "new_post_id": new_post_id,
                        "success": true,
                    });
                    format_and_output(&[result], ctx.output, None);
                }
                None => {
                    anyhow::bail!("Reply appeared to succeed but no post ID returned");
                }
            }
        }

        crate::ForumsCommands::Delete { post_id } => {
            let ok = delete_post_api(&ctx.client, &ctx.session, *post_id).await?;
            if ok {
                ctx.log.success(&format!("Post {} deleted successfully!", post_id));
                let result = serde_json::json!({
                    "action": "delete",
                    "post_id": *post_id,
                    "success": true,
                });
                format_and_output(&[result], ctx.output, None);
            } else {
                anyhow::bail!("Failed to delete post {}", post_id);
            }
        }
    }

    Ok(())
}

async fn list_forums(ctx: &ApiCtx, classification: &str) -> Result<()> {
    let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, classification).await?;
    let course_ids: Vec<u64> = courses.iter().map(|c| c.id).collect();
    let forums = get_forums_api(&ctx.client, &ctx.session, &course_ids).await?;

    let course_map: std::collections::HashMap<u64, &str> = courses.iter()
        .map(|c| (c.id, c.fullname.as_str()))
        .collect();

    let items: Vec<serde_json::Value> = forums.iter().filter_map(|f| {
        let course_id = f.get("course")
            .or_else(|| f.get("courseid"))
            .and_then(|v| v.as_u64())?;
        let course_name = course_map.get(&course_id).copied().unwrap_or("Unknown");
        Some(serde_json::json!({
            "course_id": course_id,
            "course_name": course_name,
            "cmid": f.get("cmid"),
            "forum_id": f.get("id"),
            "name": f.get("name"),
            "intro": f.get("intro").and_then(|v| v.as_str()).map(strip_html_tags),
            "timemodified": f.get("timemodified"),
        }))
    }).collect();

    ctx.log.info(&format!(
        "Found {} courses, {} forums",
        courses.len(), items.len()
    ));
    format_and_output(&items, ctx.output, None);
    Ok(())
}
